src/HDC_HLS.c

```diff
@@ -0,0 +1,140 @@
+#include "HDC_HLS.h"
+#include <math.h>
+
+/* Encodes a sample vector by multiplying it with the basis matrix */
+/*min = DIIMENSION, max = DIMENSION *-1*/
+void encode(FeatType sample[FEATURES], FeatType basis[DIMENSIONS*FEATURES],
+            FeatType encoded[DIMENSIONS], FeatType* min, FeatType* max){
+
+	#pragma HLS INTERFACE m_axi port=sample offset=slave bundle=sample
+	#pragma HLS INTERFACE s_axilite register port=sample bundle=ctrl
+	#pragma HLS INTERFACE m_axi port=basis offset=slave bundle=basis
+	#pragma HLS INTERFACE s_axilite register port=basis bundle=ctrl
+	#pragma HLS INTERFACE m_axi port=encoded offset=slave bundle=encoded
+	#pragma HLS INTERFACE s_axilite register port=encoded bundle=ctrl
+	#pragma HLS INTERFACE s_axilite register port=return bundle=ctrl
+
+    FeatType sample_local[FEATURES];
+    FeatType basis_local[DIMENSIONS*FEATURES];
+    FeatType encoded_local[DIMENSIONS];
+
+	/*
+	Partion the basis_local into blocks of size Dimensions.
+	It is bc the array can be too big to load in at once
+	*/
+	#pragma HLS ARRAY_PARTITION variable=basis_local type=block factor=DIMENSIONS
+
+    for (int i = 0; i < FEATURES; i++) {
+		#pragma HLS pipeline
+        sample_local[i] = sample[i];
+    }
+
+    for (int i = 0; i < FEATURES * DIMENSIONS; i++) {
+		#pragma HLS pipeline
+        basis_local[i] = basis[i];
+    }
+
+    //matrixmult(sample_local, basis_local, *min, *max, encoded_local, min, max);
+    matrixmult(sample_local, basis_local, encoded_local, min, max);
+    mapper(encoded_local, min, max);
+    for (int i = 0; i < DIMENSIONS; i++) {
+		#pragma HLS pipeline
+        encoded[i] = encoded_local[i];
+    }
+
+}
+
+/* Does the matrix multiplication for the encode step
+ * min max are the input and read only
+ * minR.and maxR are the output and write only
+ */
+void matrixmult(//input
```

Comment:
I don't really like putting input and output comments within the parameter list. Can we put it in the above comment instead?